lib/Backend/BailOut.cpp

```diff
@@ -13,11 +13,82 @@
 
+// In `FillBailOutRecord`, some of the fields of BailOutInfo are modified directly,
+// so simply doing a shallow copy of pointers when duplicating the BailOutInfo to
+// the helper calls for lazy bailouts will mess things up.Make a deep copies of such fields.
+void BailOutInfo::PartialDeepCopyTo(BailOutInfo * const other) const
+{
+    // Primitive types
+    other->wasCloned = this->wasCloned;
+    other->isInvertedBranch = this->isInvertedBranch;
+    other->sharedBailOutKind = this->sharedBailOutKind;
+    other->isLoopTopBailOutInfo = this->isLoopTopBailOutInfo;
+    other->bailOutOffset = this->bailOutOffset;
+    other->polymorphicCacheIndex = this->polymorphicCacheIndex;
+    other->startCallCount = this->startCallCount;
+    other->totalOutParamCount = this->totalOutParamCount;
+    other->stackLiteralBailOutInfoCount = this->stackLiteralBailOutInfoCount;
+#if DBG
+    other->wasCopied = this->wasCopied;
+#endif
+
+    other->bailOutRecord = this->bailOutRecord;
+
+    this->capturedValues->CopyTo(this->bailOutFunc->m_alloc, other->capturedValues);
+    this->usedCapturedValues->CopyTo(this->bailOutFunc->m_alloc, other->usedCapturedValues);
+
+    if (this->byteCodeUpwardExposedUsed != nullptr)
+    {
+        other->byteCodeUpwardExposedUsed = this->byteCodeUpwardExposedUsed->CopyNew(this->bailOutFunc->m_alloc);
+    }
+
+    if (this->liveVarSyms != nullptr)
+    {
+        other->liveVarSyms = this->liveVarSyms->CopyNew(this->bailOutFunc->m_alloc);
+    }
+
+    if (this->liveLosslessInt32Syms != nullptr)
+    {
+        other->liveLosslessInt32Syms = this->liveLosslessInt32Syms->CopyNew(this->bailOutFunc->m_alloc);
+    }
+
+    if (this->liveFloat64Syms != nullptr)
+    {
+        other->liveFloat64Syms = this->liveFloat64Syms->CopyNew(this->bailOutFunc->m_alloc);
+    }
+
+    if (this->outParamInlinedArgSlot != nullptr)
+    {
+        other->outParamInlinedArgSlot = this->outParamInlinedArgSlot->CopyNew(this->bailOutFunc->m_alloc);
+    }
+
+    other->startCallFunc = this->startCallFunc;
+    other->argOutSyms = this->argOutSyms;
+    other->startCallInfo = this->startCallInfo;
+    other->stackLiteralBailOutInfo = this->stackLiteralBailOutInfo;
+    other->outParamOffsets = this->outParamOffsets;
+
+
+#ifdef _M_IX86
+    other->outParamFrameAdjustArgSlot = this->outParamFrameAdjustArgSlot;
+    other->inlinedStartCall = this->inlinedStartCall;
+#endif
+
+    other->bailOutInstr = this->bailOutInstr;
+
+#if ENABLE_DEBUG_CONFIG_OPTIONS
+    other->bailOutOpcode = this->bailOutOpcode;
+#endif
+
+    other->bailOutFunc = this->bailOutFunc;
+    other->branchConditionOpnd = this->branchConditionOpnd;
+}
+
 void
 BailOutInfo::Clear(JitArenaAllocator * allocator)
 {
     // Currently, we don't have a case where we delete bailout info after we allocated the bailout record
-    Assert(!bailOutRecord);
+    Assert(bailOutRecord == nullptr || BailOutInfo::OnlyHasLazyBailOut(bailOutRecord->bailOutKind));
```

Comment:
>OnlyHasLazyBailOut [](start = 52, length = 18)

Please extend the above comment to capture the situation represented by this condition.